### PR TITLE
Better handling for weird data passed to the API via limit and offset

### PR DESCRIPTION
**Pull request by snipe**
So, every now and then, people think they're cute and they start passing bad data to the API to see how it breaks. Currently it will just return a 500 error, which is fine I suppose, but it tends to poop all over our logging system. 

This should set us up with sane default, while also still making sure that you're never going to get more results than the `MAX_RESULTS` in the `.env`.

When I had some debugging code in there, I tested this with multiple different URL parameters, and the numbers all look the way I'd expect:

```
[2023-04-12 11:58:32] local.DEBUG: Max Results: 200
[2023-04-12 11:58:32] local.DEBUG: Original Requested Limit: 10
[2023-04-12 11:58:32] local.DEBUG: Intval Limit: 10
[2023-04-12 11:58:32] local.DEBUG: Modified Limit: 10
[2023-04-12 11:58:32] local.DEBUG: Original Requested Offset: -2
[2023-04-12 11:58:32] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:32] local.DEBUG: Total Results: 209
[2023-04-12 11:58:32] local.DEBUG: -----------------------------
[2023-04-12 11:58:35] local.DEBUG: Max Results: 200
[2023-04-12 11:58:35] local.DEBUG: Original Requested Limit: 10
[2023-04-12 11:58:35] local.DEBUG: Intval Limit: 10
[2023-04-12 11:58:35] local.DEBUG: Modified Limit: 10
[2023-04-12 11:58:35] local.DEBUG: Original Requested Offset: -20
[2023-04-12 11:58:35] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:35] local.DEBUG: Total Results: 209
[2023-04-12 11:58:35] local.DEBUG: -----------------------------
[2023-04-12 11:58:37] local.DEBUG: Max Results: 200
[2023-04-12 11:58:37] local.DEBUG: Original Requested Limit: 100
[2023-04-12 11:58:37] local.DEBUG: Intval Limit: 100
[2023-04-12 11:58:37] local.DEBUG: Modified Limit: 100
[2023-04-12 11:58:37] local.DEBUG: Original Requested Offset: -20
[2023-04-12 11:58:37] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:37] local.DEBUG: Total Results: 209
[2023-04-12 11:58:37] local.DEBUG: -----------------------------
[2023-04-12 11:58:40] local.DEBUG: Max Results: 200
[2023-04-12 11:58:40] local.DEBUG: Original Requested Limit: 1000
[2023-04-12 11:58:40] local.DEBUG: Intval Limit: 1000
[2023-04-12 11:58:40] local.DEBUG: Modified Limit: 200
[2023-04-12 11:58:40] local.DEBUG: Original Requested Offset: -20
[2023-04-12 11:58:40] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:40] local.DEBUG: Total Results: 209
[2023-04-12 11:58:40] local.DEBUG: -----------------------------
[2023-04-12 11:58:43] local.DEBUG: Max Results: 200
[2023-04-12 11:58:43] local.DEBUG: Original Requested Limit: -1000
[2023-04-12 11:58:43] local.DEBUG: Intval Limit: -1000
[2023-04-12 11:58:43] local.DEBUG: Modified Limit: 1
[2023-04-12 11:58:43] local.DEBUG: Original Requested Offset: -20
[2023-04-12 11:58:43] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:43] local.DEBUG: Total Results: 209
[2023-04-12 11:58:43] local.DEBUG: -----------------------------
[2023-04-12 11:58:47] local.DEBUG: Max Results: 200
[2023-04-12 11:58:47] local.DEBUG: Original Requested Limit: -1000
[2023-04-12 11:58:47] local.DEBUG: Intval Limit: -1000
[2023-04-12 11:58:47] local.DEBUG: Modified Limit: 1
[2023-04-12 11:58:47] local.DEBUG: Original Requested Offset: farts
[2023-04-12 11:58:47] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:47] local.DEBUG: Total Results: 209
[2023-04-12 11:58:47] local.DEBUG: -----------------------------
[2023-04-12 11:58:53] local.DEBUG: Max Results: 200
[2023-04-12 11:58:53] local.DEBUG: Original Requested Limit: -5) OR 584=(SELECT 584 FROM PG_SLEEP(15))--
[2023-04-12 11:58:53] local.DEBUG: Intval Limit: -5
[2023-04-12 11:58:53] local.DEBUG: Modified Limit: 1
[2023-04-12 11:58:53] local.DEBUG: Original Requested Offset: 20
[2023-04-12 11:58:53] local.DEBUG: Modified Offset: 20
[2023-04-12 11:58:53] local.DEBUG: Total Results: 209
[2023-04-12 11:58:53] local.DEBUG: -----------------------------
[2023-04-12 11:59:15] local.DEBUG: Max Results: 200
[2023-04-12 11:59:15] local.DEBUG: Original Requested Limit: -5) OR 584=(SELECT 584 FROM PG_SLEEP(15))--
[2023-04-12 11:59:15] local.DEBUG: Intval Limit: -5
[2023-04-12 11:59:15] local.DEBUG: Modified Limit: 1
[2023-04-12 11:59:15] local.DEBUG: Original Requested Offset: farts
[2023-04-12 11:59:15] local.DEBUG: Modified Offset: 0
[2023-04-12 11:59:15] local.DEBUG: Total Results: 209
[2023-04-12 11:59:15] local.DEBUG: -----------------------------
```

I've tested this and all looks well, however this touches basically every major listing endpoint we have, so I'd appreciate some additional testing :)

### CODE REVIEW

**Review by uberbrady — COMMENTED** [review]
This seems pretty reasonable, but I do have a couple of questions - 

- We don't use the `intval()` thing when comparing the `$request->input('limit')` value to the `config('app.max_results')` value - I don't know if that's a big deal or not though.
- We are doing a second `max()` call when we already compared these values before. Is that necessary?
- All of this feels a little...repetitive. Is there something we can do that's just a little more systemic so that we aren't repeating ourselves all the time? Maybe something from here: https://laravel.com/docs/9.x/pagination ?

All of that being said, those are just tiny nits and I don't think sufficient to block this from acceptance, if I can do a couple of key tests first (which, I will do).

Please let me know your thoughts on this and I'll get to testing.

**Comment by marcusmoore** [comment]
> We don't use the intval() thing when comparing the $request->input('limit') value to the config('app.max_results') value - I don't know if that's a big deal or not though.

@uberbrady I think we can safely assume `config('app.max_results')` is an integer since it pulls `MAX_RESULTS` from the environment or defaults to an integer. It wouldn't hurt to wrap it in `intval()` but I don't think it's necessary.

File: app/Http/Controllers/Api/AccessoriesController.php [hunk 1]
```diff
@@ -80,12 +80,9 @@ public function index(Request $request)
...
 
-        // Set the offset to the API call's offset, unless the offset is higher than the actual count of items in which
-        // case we override with the actual count, so we should return 0 items.
-        $offset = (($accessories) && ($request->get('offset') > $accessories->count())) ? $accessories->count() : $request->get('offset', 0);
-
-        // Check to make sure the limit is not higher than the max allowed
-        ((config('app.max_results') >= $request->input('limit')) && ($request->filled('limit'))) ? $limit = $request->input('limit') : $limit = config('app.max_results');
+        // Make sure the offset and limit are actually integers and do not exceed system limits
+        $offset = ($request->input('offset') > $accessories->count()) ? $accessories->count() : intval(request('offset'));
+        $limit = ($request->input('limit') > config('app.max_results')) ? config('app.max_results') : max(intval(request('offset')),  config('app.max_results'));
```
Comment (marcusmoore) [inline app/Http/Controllers/Api/AccessoriesController.php:85]: I might be missing something but let's say `MAX_RESULTS=50` and the user requests `?limit=10` then `$limit` would be set to `50`. Is that expected?

Comment (snipe) [reply app/Http/Controllers/Api/AccessoriesController.php:85]: I don't *think* so?

```
[2023-04-12 11:58:32] local.DEBUG: Max Results: 200
[2023-04-12 11:58:32] local.DEBUG: Original Requested Limit: 10
[2023-04-12 11:58:32] local.DEBUG: Intval Limit: 10
[2023-04-12 11:58:32] local.DEBUG: Modified Limit: 10
[2023-04-12 11:58:32] local.DEBUG: Original Requested Offset: -2
[2023-04-12 11:58:32] local.DEBUG: Modified Offset: 0
[2023-04-12 11:58:32] local.DEBUG: Total Results: 209
```

I can dig deeper shortly, but I believe it's doing the right thing here.

**Comment by snipe** [comment]
@uberbrady 

> We don't use the intval() thing when comparing the $request->input('limit') value to the config('app.max_results') value - I don't know if that's a big deal or not though.

I don't think we really need it there - if you pass us a real number, it wouldn't matter. If you pass us garbage, that will shake out when we finally decide what value it needs to be.

> We are doing a second max() call when we already compared these values before. Is that necessary?

I'm not sure what you mean. 

> All of this feels a little...repetitive. Is there something we can do that's just a little more systemic so that we aren't repeating ourselves all the time? Maybe something from here: https://laravel.com/docs/9.x/pagination ?

As discussed on the call this AM, yes and no. Bootstrap-table requires our data to exist in a specific format (which is why we have all of those API transformers, etc.) It's possible we can get the best of both worlds - it's something I've been thinking about for a while. This PR is just meant to be a middle-step so shitheels who bought Burpsuite for Dummies don't poop all over our logs, but it's very possible that there's a better longer term solution possible here.

<blockquote><div><strong><a href="https://laravel.com/">Laravel - The PHP Framework For Web Artisans</a></strong></div><div>Laravel is a PHP web application framework with expressive, elegant syntax. We’ve already laid the foundation — freeing you to create without sweating the small things.</div></blockquote>

**Comment by snipe** [comment]
An alternative to this PR would be a piece of middleware (or in a provider or error handler) that simply rejects `limit` and `offset` in any request if their values are not greater than 0 and are not numbers.

**Comment by marcusmoore** [comment]
> An alternative to this PR would be a piece of middleware (or in a provider or error handler) that simply rejects limit and offset if their values are not greater than 0 and are not numbers.

Hm...I like that idea a lot...

**Comment by snipe** [comment]
I'm going to merge this into develop now and keep testing

**Pull request merged**